crates/red_knot_python_semantic/src/types.rs

```diff
@@ -88,9 +88,64 @@ fn symbol_ty<'db>(db: &'db dyn Db, scope: ScopeId<'db>, name: &str) -> Type<'db>
         .unwrap_or(Type::Unbound)
 }
 
-/// Shorthand for `symbol_ty` that looks up a module-global symbol by name in a file.
+/// Return a list of the symbols that typeshed declares in the scope of `types.ModuleType`
+#[salsa::tracked(return_ref)]
+fn module_type_symbols<'db>(db: &'db dyn Db) -> Box<[Box<str>]> {
+    let Some(module_type) = KnownClass::ModuleType
+        .to_class(db)
+        .into_class_literal_type()
+    else {
+        return Box::default();
+    };
+
+    let module_type_scope = module_type.body_scope(db);
+    let module_type_symbol_table = symbol_table(db, module_type_scope);
+
+    // `__dict__` and `__init__` are very special members that can be accessed as attributes
+    // on the module when imported, but cannot be accessed as globals *inside* the module.
+    //
+    // `__getattr__` is even more special: it doesn't exist at runtime, but typeshed includes it
+    // to reduce false positives associated with functions that dynamically import modules
+    // and return `Instance(types.ModuleType)`. We should ignore it for any known module-literal type.
+    let module_type_symbols: Box<[Box<str>]> = module_type_symbol_table
+        .symbols()
+        .filter(|symbol| symbol.is_declared())
+        .map(|symbol| &**symbol.name())
+        .filter(|symbol_name| !matches!(*symbol_name, "__dict__" | "__getattr__" | "__init__"))
+        .map(Box::from)
+        .collect();
+
+    // Make sure it contains declared symbols in the `types.ModuleType` scope,
+    // but not other random symbols that are only *referenced* in the `types.ModuleType` scope
+    debug_assert!(module_type_symbols.contains(&Box::from("__name__")));
+    debug_assert!(!module_type_symbols.contains(&Box::from("property")));
```

Comment:
These assertions feel more like a test than a debug assertion to me.

Reply:
That's fair. I couldn't see a way of doing an mdtest asserting this, but I suppose I can still just write an old-fashioned unittest for the function!